Release notes for v3.0.0-preview3 (#1079)
dotnet/SqlClient: release-notes/3.0/3.0.0-preview3.md

# Release Notes

## Microsoft.Data.SqlClient 3.0.0-preview3.*****.* released 20 May 2021

This update brings the below changes over the previous release:

### Added

> - Added support for "Active Directory Default" authentication mode [#1043](https://github.com/dotnet/SqlClient/pull/1043) [Read more](#active-directory-default-authentication)

Reviewer: ```suggestion
- Added support for "Active Directory Default" authentication mode [#1043](https://github.com/dotnet/SqlClient/pull/1043) [Read more](#active-directory-default-authentication-support)
```

> Further customization options are not available at the moment.

### Custom master key store provider registration enhancements

Microsoft.Data.SqlClient now offers more control of where master key store providers are accessible in an application in order to better support multi-tenant applications and their use of column encryption/decryption. The following APIs are introduced to allow registration of custom master key store providers on instances of `SqlConnection` and `SqlCommand`:

```cs
public class SqlConnection
{
        public void RegisterColumnEncryptionKeyStoreProvidersOnConnection(IDictionary<string, SqlColumnEncryptionKeyStoreProvider> customProviders)
}
public class SqlCommand 
{
        public void RegisterColumnEncryptionKeyStoreProvidersOnCommand(IDictionary<string, SqlColumnEncryptionKeyStoreProvider> customProviders)
}
```

The static API on `SqlConnection`, i.e. `SqlConnection.RegisterColumnEncryptionKeyStoreProviders` to register custom master key store providers globally continues to be supported. The column encryption key cache maintained globally only applies to globally registered providers.

#### Column master key store provider registration precedence

> The built-in column master key store providers that are available for the Windows Certificate Store, CNG Store and CSP are pre-registered. No providers should be registered on the connection or command instances if one of the built-in column master key store provider is needed.

Reviewer 2: ```suggestion
The built-in column master key store providers that are available for the Windows Certificate Store, CNG Store and CSP are pre-registered. No providers should be registered on the connection or command instances if one of the built-in column master key store providers is needed.
```

> System key store providers will be used from the global system cache if the requested provider is one. No providers should be registered on the connection or command instances if a system provider is needed.
> Registered key store providers are cached by the driver in 3 different cache layers. The global layer is as-is. The new per-connection and per-command level caches will be empty initially and can be set more than once.
> 
> The precedence of the three caches will be as follows:
> 
> - The per-command cache will be checked if it is not empty.
> - If the per-command cache is empty, the per-connection cache will be checked if it is not empty.
> - If the per-connection cache is empty, the global cache will be checked.
> 
> Once a cache is found, the driver will **NOT** fall back to the other caches to search for requested provider.
> An exception will be thrown containing only the registered providers in the cache that was checked.

Reviewer 2: What exactly is a "system" provider? I find the terms global and system confusing without some clarification. I assume "system" refers to providers for which there is only one possible store on the computer running the program.

Here's my suggestion, but I may have some misunderstandings:
```suggestion
System key store providers are available for the Windows Certificate Store and ???. They will be used from the global registration if the requested provider is a system provider. No providers should be registered on the connection or command instances if a system provider is needed.
Key store providers can be registered with the driver at three different layers. The global level is as it currently is. The new per-connection and per-command level registrations will be empty initially and can be set more than once.

The precedence of the three registrations are as follows:

- The per-command registration will be checked if it is not empty.
- If the per-command registration is empty, the per-connection registration will be checked if it is not empty.
- If the per-connection registration is empty, the global registration will be checked.

Once any key store provider is found at a registration level, the driver will **NOT** fall back to the other registrations to search for a provider.
If providers are registered but the proper provider is not found at a level, an exception will be thrown containing only the registered providers in the registration that was checked.
```

Author: Yes, System Providers are Windows-only, available via application host OS only. I'll change the term to "built-in column master key store providers" (as per MS docs)

#### Column Encryption Key cache precedence

> The column encryption keys (CEKs) for custom key store providers registered using the new instance-level APIs will not be cached by the driver. The key store providers shall implement their own cache to gain performance. This local cache of column encryption keys implemented by custom key store providers will be disabled by the driver if the key store provider instance is registered in the driver globally.

Reviewer 2: ```suggestion
The column encryption keys (CEKs) for custom key store providers registered using the new instance-level APIs will not be cached by the driver. The key store providers need to implement their own cache to gain performance. This local cache of column encryption keys implemented by custom key store providers will be disabled by the driver if the key store provider instance is registered in the driver at the global level.
```

A new API has also been introduced on the `SqlColumnEncryptionKeyStoreProvider` base class to set the cache time to live:

```cs
public abstract class SqlColumnEncryptionKeyStoreProvider
{
    // The default value of Column Encryption Key Cache Time to Live is 0.
    // Provider's local cache is disabled for globally registered providers.
    // Custom key store provider implementation must include column encryption key cache to provide caching support to locally registered providers.
    public virtual TimeSpan? ColumnEncryptionKeyCacheTtl { get; set; } = new TimeSpan(0);
}
```

### IP Address preference

> A new connection property `IPAddressPreference` is introduced to specify the IP address family preference to the driver when establishing TCP connection. Below are the three accepted values for this property:

Reviewer 2: ```suggestion
A new connection property `IPAddressPreference` is introduced to specify the IP address family preference to the driver when establishing TCP connections. Below are the three accepted values for this property:
```

- **IPv4First**

>   - This is the default preference value. The driver will traverse IPv4 addresses first. If none of them can be connected successfully, it continues to try IPv6 addresses if there are any.

Reviewer 2: ```suggestion
  - This is the default preference value. The driver will use resolved IPv4 addresses first. If none of them can be connected to successfully, it will try resolved IPv6 addresses.
```

- **IPv6First**

>   - The driver will traverse IPv6 addresses first. If none of them can be connected successfully, it continues to try IPv4 addresses if there are any.

Reviewer 2: ```suggestion
  - The driver will use resolved IPv6 addresses first. If none of them can be connected to successfully, it will try resolved IPv4 addresses.
```

> A new connection property `IPAddressPreference` is introduced to specify the IP address family preference to the driver when establishing TCP connections. Below are the three accepted values for this property:

Reviewer: Could you add a note about its conflict with `Multi Subnet Failover` and `Transparent Network IP Resolution` settings in .NET Framework and .NET Core.

Author: ```suggestion
A new connection property `IPAddressPreference` is introduced to specify the IP address family preference to the driver when establishing TCP connections. If `Transparent Network IP Resolution` (in .NET Framework) or `Multi Subnet Failover` is set to `true`, this setting has no effect. Below are the three accepted values for this property:
```

- **UsePlatformDefault**

>   - The driver will traverse all the IP addresses in the same order as received from DNS resolution response.

Reviewer 2: ```suggestion
  - The driver will try IP addresses in the order received from the DNS resolution response.
```
